functions/slack/search.go

```diff
@@ -71,27 +94,85 @@ func KGSearch(w http.ResponseWriter, r *http.Request) {
 
 // [END functions_slack_search]
 
+// [START functions_slack_request]
+func makeSearchRequest(query string) (*Message, error) {
+	res, err := entitiesService.Search().Query(query).Limit(1).Do()
+	if err != nil {
+		return nil, fmt.Errorf("Do: %v", err)
+	}
+	return formatSlackMessage(query, res)
+}
+
+// [END functions_slack_request]
+
 // [START functions_verify_webhook]
-func verifyWebHook(form url.Values) error {
-	t := form.Get("token")
-	if len(t) == 0 {
-		return fmt.Errorf("empty form token")
+
+// VeryfyWebhook now uses signature verification instead of tokens.
+// see https://api.slack.com/docs/verifying-requests-from-slack
+func verifyWebHook(r *http.Request, secret string) error {
+	result, err := verifyRequestSignature(r, secret)
+
+	if err != nil {
+		return fmt.Errorf("verifyWebHook: %v", err)
 	}
-	if t != config.Token {
-		return fmt.Errorf("invalid request/credentials: %q", t[0])
+
+	if !result {
+		return fmt.Errorf("invalid request/credentials")
 	}
+
 	return nil
 }
 
-// [END functions_verify_webhook]
+func getSignature(base []byte, secret []byte) []byte {
+	h := hmac.New(sha256.New, secret)
+	h.Write(base)
 
-// [START functions_slack_request]
-func makeSearchRequest(query string) (*Message, error) {
-	res, err := entitiesService.Search().Query(query).Limit(1).Do()
+	return h.Sum(nil)
+}
+
+// Arbitrarily trusting requests time stamped less than 5 minutes ago.
+func checkTimestamp(timeStamp int64) (bool, time.Duration) {
+	t := time.Since(time.Unix(timeStamp, 0))
+
+	return t.Minutes() <= 5, t
+}
+
+func verifyRequestSignature(r *http.Request, slackSigningSecret string) (bool, error) {
+	timeStamp := r.Header.Get(slackRequestTimestampHeader)
+	slackSignature := r.Header.Get(slackSignatureHeader)
+
+	t, err := strconv.ParseInt(timeStamp, 10, 64)
 	if err != nil {
-		return nil, fmt.Errorf("Do: %v", err)
+		return false, fmt.Errorf("error: %v | got timeStamp: %s", err, timeStamp)
```

Comment:
No need to include "error" in the error message, since it's already an `error`. Could change to something like:

```go
return false, fmt.Errorf("strconv.ParseInt(%s): %v", timeStamp, err)
```

This follows the pattern for test failures mentioned [here](https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures), but also makes for a useful error message.

Same with "error" in the other errors below.